File: lib/python/opentoken/src/main/opentoken/attributes/base_attribute.py
from typing import List
from opentoken.attributes.serializable_attribute import SerializableAttribute
from opentoken.attributes.validation.not_null_or_empty_validator import NotNullOrEmptyValidator
from opentoken.attributes.validation.serializable_attribute_validator import SerializableAttributeValidator


class BaseAttribute(SerializableAttribute):
    """A base implementation of the SerializableAttribute interface.

    This class provides a default implementation of the validate method
    that validates the attribute value against a set of validation rules.

    The default validation rules are:
    - Not null or empty
    """

    def __init__(self, validation_rules: List[SerializableAttributeValidator] = None):
        if validation_rules is None:
            validation_rules = []

        rule_list = [NotNullOrEmptyValidator()]
        rule_list.extend(validation_rules)
        self.validation_rules = rule_list

    def validate(self, value: str) -> bool:
        """Validates the attribute value against a set of validation rules."""
        if value is None:
            return False

        return all(rule.eval(value) for rule in self.validation_rules)

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/BaseAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes;

import java.util.ArrayList;
import java.util.List;
import com.truveta.opentoken.attributes.validation.NotNullOrEmptyValidator;
import com.truveta.opentoken.attributes.validation.SerializableAttributeValidator;

/**
 * A base implementation of the {@link SerializableAttribute} interface.
 * 
 * <p>
 * This class provides a default implementation of the
 * {@link Attribute#validate(String)}
 * method that validates the attribute value against a set of validation rules.
 * </p>
 * 
 * <p>
 * The default validation rules are:
 * <ul>
 * <li>Not null or empty</li>
 * </ul>
 * </p>
 */
public abstract class BaseAttribute implements SerializableAttribute {

    private static final long serialVersionUID = 1L;

    private final List<SerializableAttributeValidator> validationRules;

    protected BaseAttribute(List<SerializableAttributeValidator> validationRules) {
        ArrayList<SerializableAttributeValidator> ruleList = new ArrayList<>();
        ruleList.add(new NotNullOrEmptyValidator());
        ruleList.addAll(validationRules);
        this.validationRules = List.copyOf(ruleList);
    }

    /**
     * Validates the attribute value against a set of validation rules.
     */
    @Override
    public boolean validate(String value) {
        if (value == null) {
            return false;
        }

        return validationRules.stream().allMatch(rule -> rule.eval(value));
    }
}
